### DISCUSSION TITLE
CRC and CheckSum ERROR, N64, GB, SNES (CLOSED, MAINBOARD DEFECT)

### DISCUSSION THREAD

**Discussion by mauruco** [sanni/cartreader#210] [Need help]
Hello guys. I wanted to take the opportunity to give thanks for the beautiful project that is SanniCartReader.
I have a problem, all the games I tried to read are giving checksum error or crc error.
I already formatted the SD cart and copied the files to root, tried FAT32 and exFAT.
Any help is welcome it's my first SanniCartReader and I'm kind of lost.
1. What could be the problem with SD Cart reader?
2. Taking the opportunity, how can I test if the clock generator is working correctly?

**Comment by sanni** [comment]
You can test the clock generator like this: https://forum.arduino.cc/t/rom-reader-for-super-nintendo-super-famicom-game-cartridges/155260/1143#msg4518921

Make sure you clean the cartridge contacts with a q-tip and rubbing alcohol until the q-tip does not experience any black discoloration anymore. Especially SNES and GBA carts are prone to read errors with the Cart Reader when even slightly dirty.

If after the cleaning procedure you still get read errors then you most likely have a bad solder joint or a short circuit somewhere on the Cart Reader PCB, the Arduino or the SD card module.

You can guess where the error is by comparing a bad dump with a known good dump in an hex editor. Depending on the pattern the wrong bytes have it might hint at a specific data or address line which will make searching for the faulty connection/trace much easier.

**Comment by Dakkaron** [comment]
So reading gives you the crc errors?

Some thoughts to that:
- I honestly don't think it is an SD problem. It could be, but I wouldn't expect it to.
- Did you check all soldering connections (and resoldered them)?
- I had a similar problem that was caused by the 3D printed case of the cartreader, which had too small clearances and pushed the cart off the connections.
- Check your power supply, that it really delivers the correct voltage.
- Reproduction cartridges with a "save to flash" feature always return bad crc

**Comment by mauruco** [comment]
I just tried:
  1. Reflow, with Amtech 559 all soldering points, looks okay.
  2. I replaced the clock generator with another one.
  3. I replaced the SD module with another one.
Even so the error persists. All my cartridges are genuine and very clean.

HEX:
   I opened a dump with the hex editor, in the first line there are differences.
   N64: My Duke Nukem dump has "DECE NUKEM" in the header, a good dump has "DUKE NUKEN".
   N64: My Majora's Maks dump has the header "REDDA EABORA'C", a good dump "ZELDA MAJORA'S MASK".
   SNES: All my dumps don't match anything I can find.
   GB: My CartReader doesn't even save gameboy dumps to SD (I only have one game, Wario Land).
   
I'm waiting for a new SD Card (Sandisk Ultra 64GB CAT 10) to test.
Any more ideas? Should I try another version of the software on arduino?
Thank you for your help.

**Reply by Dakkaron** [reply]
If you think it's an SD problem, go for an older model.
Advances in newer SDs are wasted on this anyways (the Arduino and the cartridges are ancient and very slow. Even the slowest SD card you can find is faster than them; also, size does not matter, since ROMs are so small).
For compatibility it is often best to take a 2GB card, since that's the maximum size of the original SD spec and has the best compatibility.

**Reply by mauruco** [reply]
I've tried 5 different ones and the error persists.

**Comment by sanni** [comment]
Can you please use a multimeter in diode test mode, where it peeps if two traces are connected to each other, and check if those two points are connected to each other.

![gewrgehsrgeghe](https://user-images.githubusercontent.com/19169494/127720205-e8101685-0f6a-408a-85fe-51bd0cb92a42.jpg)

Those are Analog Pin 10 and 11 on the Arduino Mega.

**Reply by mauruco** [reply]
I took out the arduino and tested the two points on the motherboard, they don't make contact.
I tested pins 10 and 11 of the arduino and they don't make contact.
I connected the arduino again with the motherboard and the pins don't make contact.
I tested if all arduino pins are making contact with the motherboard, yes, they are.
Taking advantage of this, I followed the red line marked on the image, from pin 11, it is the same as the image.
Sorry for my translator english and thanks for the help.
Could it be a defective Arduino?

**Reply by majorpbx** [reply]
Forgive me if this has been discussed but have you tried testing from the socket connector pin to the arduino pin? I ask this as my problem with roms not reading correctly stemmed from the leads inside the socket itself not making adequate contact with the cartridge. I'm by no means an expert but I did chase my tail with a potential bad trace when I never found an issue. It was when I got the idea to test from the cartridge board to the soldered side of the socket that I found my real issue. Just a thought.

**Reply by Dakkaron** [reply]
I had a very similar problem with the cartridge not making cood contact (it was because the printed case didn't have enough clearance and pushed the cart sideways.

Another way to diagnose that is to lightly pust the cartridge in any direction while flashing. If the errors disappear, that's your problem.

**Reply by sanni** [reply]
That is a good point. 

Here's an idea. Without any cartridge inserted pretend to dump a 4MB N64 ROM:
Nintendo 64 -> Game Cartridge -> Right button to ignore Cartridge Unknown error -> 4MB -> Read Rom -> Ignore

It should create a GPERROR folder with a 4MB file inside. Open this file in a hex editor. It should have this weird pattern that slowly counts up from 0x0 in steps of 2. If there was a short between two pins it would look different I think. 

Since this file will not contain any copyrighted material you can also zip it and upload it here. Then I can compare it to my file.

![wqtgewhwhehrhr](https://user-images.githubusercontent.com/19169494/127737104-8505bc92-656e-4d08-8f1d-6f2e6b6074fe.jpg)

(Unrelated to this issue but I think that this pattern in itself is a bug in the N64 code because the AD pins are still holding the bits from the address when it reads the data. Might be worth investigating further. The GBA code does not experience this behavior.)

**Reply by mauruco** [reply]
@sanni

**Reply by mauruco** [reply]
@Dakkaron, I haven't made the case yet.

**Reply by mauruco** [reply]
@majorpbx, I think I'll have to test all the pins. Whenever I've tested a pin, I've tried to test it end-to-end, but I might have done it wrong.

**Comment by mauruco** [comment]
I have a little oscilloscope here, I can try to get some kind of measurement if that's a good idea.
SNES (Read Rom Menu) without connected cartridge has on channel 0 a very fast wave for my equipment, channel 1 nothing, channel 2 4.35MGHz - 4.41MGHz.
N64 Read Rom Menu) without connected cartridge has on channel 0 nothing, channel 1 nothing 2.83 MGHz - 2.86 MGHz, channel 2 nothing.
I took the measurements of the points on the motherboard where the wave generator is connected.

**Comment by mauruco** [comment]
Please close, motherboard is defective.
Thank you all

**Comment by Vek001** [comment]
mauruco
I have same problem,
In which part of the motherboard did you find the defect?

**Reply by mauruco** [reply]
Hi, it wasn't just one place. The board was leaking green paint and had some redone dots. When I saw this, I quit immediately, because I had others boards to start from scratch.
I would look at each soldering point in your case with a good magnifying glass and lots of light. my second and third attempts were successful.
you can also compare your rom with a good dump, opening it in a hexeditor, so it might be possible to tell you the area on the board to check it out better.

**Reply by Vek001** [reply]
Hi mauruco
Thank you for re-mail
I had 3/10 maked. 33%...
I will chack agein 
thank you
